File: 3rd-semester/object-oriented-programming/1st-assignment/java/src/main/java/com/university/oop/surveillance/model/Suspect.java
package com.university.oop.surveillance.model;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Objects;
import java.util.logging.Logger;

/**
 * Represents a surveillance suspect with associated phone numbers,
 * known collaborators, and location information.
 */
public class Suspect {

    private static final Logger LOGGER = Logger.getLogger(Suspect.class.getName());

    private String name;
    private String codeName;
    private String country;
    private String city;

    private final List<String>  phoneNumbers  = new ArrayList<>();
    private final List<Suspect> collaborators = new ArrayList<>();

    public Suspect() {}

    /**
     * Constructs a Suspect with full identity information.
     *
     * @param name     real name
     * @param codeName operational alias
     * @param country  country of origin
     * @param city     city of residence
     */
    public Suspect(String name, String codeName, String country, String city) {
        this.name     = name;
        this.codeName = codeName;
        this.country  = country;
        this.city     = city;
    }

    public String getName()     { return name; }
    public void setName(String name) { this.name = name; }

    public String getCodeName() { return codeName; }
    public void setCodeName(String codeName) { this.codeName = codeName; }

    public String getCountry()  { return country; }
    public void setCountry(String country) { this.country = country; }

    public String getCity()     { return city; }
    public void setCity(String city) { this.city = city; }

    /** @return unmodifiable view of phone numbers */
    public List<String> getPhoneNumbers() {
        return Collections.unmodifiableList(phoneNumbers);
    }

    /** @return unmodifiable view of known collaborators */
    public List<Suspect> getCollaborators() {
        return Collections.unmodifiableList(collaborators);
    }

    /** @return number of known collaborators */
    public int getCollaboratorCount() { return collaborators.size(); }

    /**
     * Registers a phone number. Duplicates are silently ignored.
     *
     * @param number phone number to add
     */
    public void addPhoneNumber(String number) {
        if (!phoneNumbers.contains(number)) {
            phoneNumbers.add(number);
        } else {
            LOGGER.warning("Phone number already registered: " + number);
        }
    }

    /**
     * Adds a collaborator. Duplicates are silently ignored.
     *
     * @param suspect the collaborator to add
     */
    public void addCollaborator(Suspect suspect) {
        if (!collaborators.contains(suspect)) {
            collaborators.add(suspect);
        }
    }

    /**
     * Checks whether this suspect is directly linked to another.
     *
     * @param suspect the suspect to check
     * @return {@code true} if connected
     */
    public boolean isConnectedTo(Suspect suspect) {
        return collaborators.contains(suspect);
    }

    /**
     * Returns suspects who are collaborators of both this suspect and another.
     *
     * @param other the other suspect
     * @return list of common partners (computed fresh each call)
     */
    public List<Suspect> getCommonPartners(Suspect other) {
        List<Suspect> common = new ArrayList<>();
        for (Suspect s : collaborators) {
            if (other.getCollaborators().contains(s)) {
                common.add(s);
            }
        }
        return common;
    }

    /**
     * Logs all collaborators; marks those from the same country with {@code *}.
     */
    public void logCollaborators() {
        for (Suspect s : collaborators) {
            String marker = country.equals(s.getCountry()) ? "*" : "";
            LOGGER.info(String.format("Name: %s | CodeName: %s%s",
                    s.getName(), s.getCodeName(), marker));
        }
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Suspect other)) return false;
        return Objects.equals(name, other.name) && Objects.equals(codeName, other.codeName);
    }

    @Override
    public int hashCode() { return Objects.hash(name, codeName); }

    @Override
    public String toString() {
        return String.format("Suspect[name=%s, codeName=%s, country=%s, city=%s]",
                name, codeName, country, city);
    }
}

File: 3rd-semester/object-oriented-programming/1st-assignment/csharp/src/Models/Suspect.cs
namespace SurveillanceRegistry.Models;

/// <summary>Represents a surveillance suspect with phone numbers, collaborators, and location data.</summary>
public class Suspect
{
    private readonly List<string>  _phoneNumbers  = [];
    private readonly List<Suspect> _collaborators = [];

    public string Name     { get; set; } = string.Empty;
    public string CodeName { get; set; } = string.Empty;
    public string Country  { get; set; } = string.Empty;
    public string City     { get; set; } = string.Empty;

    public IReadOnlyList<string>  PhoneNumbers  => _phoneNumbers.AsReadOnly();
    public IReadOnlyList<Suspect> Collaborators => _collaborators.AsReadOnly();
    public int CollaboratorCount => _collaborators.Count;

    public Suspect() { }
    public Suspect(string name, string codeName, string country, string city)
    { Name = name; CodeName = codeName; Country = country; City = city; }

    public void AddPhoneNumber(string number)
    { if (!_phoneNumbers.Contains(number)) _phoneNumbers.Add(number); }

    public void AddCollaborator(Suspect suspect)
    { if (!_collaborators.Contains(suspect)) _collaborators.Add(suspect); }

    public bool IsConnectedTo(Suspect suspect) => _collaborators.Contains(suspect);

    public IReadOnlyList<Suspect> GetCommonPartners(Suspect other) =>
        _collaborators.Where(s => other.Collaborators.Contains(s)).ToList();

    public override string ToString() =>
        $"Suspect[name={Name}, codeName={CodeName}, country={Country}, city={City}]";
}
